java/arcs/core/host/SdkHandleManager.kt

```diff
@@ -0,0 +1,432 @@
+     * @property sender block used to execute callback lambdas
+     * @property idGenerator used to generate unique IDs for newly stored entities.
+     */
+    private fun <T : Entity> createSdkHandle(
+        entitySpec: EntitySpec<T>,
+        handleName: String,
+        storageHandle: StorageHandle<*, *, *>,
+        handleMode: HandleMode,
+        idGenerator: Id.Generator,
+        sender: Sender
+    ): Handle {
+        return when (storageHandle) {
+            is SingletonHandle<*> -> createSingletonHandle(
+                entitySpec,
+                handleName,
+                storageHandle as SingletonHandle<RawEntity>,
+                handleMode,
+                idGenerator,
+                sender
+            )
+            is SetHandle<*> -> createSetHandle(
+                entitySpec,
+                handleName,
+                storageHandle as SetHandle<RawEntity>,
+                handleMode,
+                idGenerator,
+                sender
+            )
+            else -> throw Exception("Unknown storage handle type ${storageHandle::class}")
+        }
+    }
+
+    /**
+     * Create a [Handle] given a [StorageHandle] and a [HandleHolder] with [EntitySpec] definitions.
+     *
+     * @property handleHolder contains handle and entitySpec declarations
+     * @property handleName name for the handle, must be present in [HandleHolder.entitySpecs]
+     * @property storageHandle a [StorageHandle] usually obtained thru [HandleManager]
+     * @property handleMode whether a handle is Read,Write, or ReadWrite (default)
+     * @property sender block used to execute callback lambdas
+     * @property idGenerator used to generate unique IDs for newly stored entities.
+     */
+    private fun createSdkHandle(
+        handleHolder: HandleHolder,
+        handleName: String,
+        storageHandle: StorageHandle<*, *, *>,
+        handleMode: HandleMode,
+        idGenerator: Id.Generator,
+        sender: Sender
+    ): Handle {
+        val entitySpec: EntitySpec<*>? =
+            handleHolder.entitySpecs[handleName] as? EntitySpec<*>
+        val handle = createSdkHandle(
+            entitySpec ?: throw EntitySpecNotFound(handleName, handleHolder),
+            handleName,
+            storageHandle,
+            handleMode,
+            idGenerator,
+            sender
+        )
+        val handleMap = handleHolder.map as MutableMap<String, Handle>
+        handleMap.put(handleName, handle)
+        return handle
+    }
+}
+
+internal open class HandleEventBase<T> {
+    protected val onUpdateActions: MutableList<(T) -> Unit> = mutableListOf()
+
+    suspend fun onUpdate(action: (T) -> Unit) {
+        onUpdateActions.add(action)
+    }
+
+    protected suspend fun fireUpdate(value: T) {
+        onUpdateActions.forEach { action -> action(value) }
+    }
+}
+
+internal open class ReadableSingletonHandleImpl<T : Entity>(
+    val entitySpec: EntitySpec<T>,
+    val handleName: String,
```

Comment:
Is there a case where we want the `handleName` of the wrapped handle to be different than that of the storage handle? If not, we could remove this property, and have the `name` property below return the `storageHandle` name.

Reply:
I'm unsure of this, Yuan reported that if two storage handles have the same name, it crashes storage service. Right now the StorageHandles use StorageKey.toString() as the name. I would say in the interests of expediency, because this code is working and not crashing, we keep it like this, and then have a meeting with Jason and others over the naming issue to get a good idea on what the Storage layer wants.